// node-red-contrib-mobisms/mobisms.js
module.exports = function(RED) {
  
    var request = require("request");

    function validPhone(inputPhone)
    {
      let validTen = /^\d{10}$/;
      let validEle = /^\d{11}$/;
      if (inputPhone.match(validTen) || inputPhone.match(validEle)) {
          return true;
      }
      else {
         return false;
      }
    }
    

    function SMSNode(config) {
        RED.nodes.createNode(this,config);
        this.name = config.name;
        this.phone = config.phone;
        this.text = config.text;
        var node = this;

        let datauser = {
            'username': 'mobifoneruleengine@mobifone.com',
            'password': 'bigdataITC'
        }

        let jsonuser = JSON.stringify(datauser)
        var tokenuser = "abc";
        node.on('input', function(msg) {
            request.post({
              headers: {'content-type' : 'application/json'},
              url: 'http://localhost:8000/rule-engine/getTokenSMS',
              body: jsonuser
            }, function(error, response, body){
              if (!error && response.statusCode == 200) {
                  var bodyjson = JSON.parse(body);
                  tokenuser = bodyjson.token;

                  //Request 2

                  if(node.text === "") {
                    node.text = msg.payload;
                  }

                  if(node.phone === "") {
                    node.warn("Required phone number!");
                  } else {
                      if(validPhone(node.phone)) {
                        let data = {
                          'isdn': node.phone,
                          'messageOut': node.text,
                          'status': 1
                        }
              
                        let jsondata = JSON.stringify(data)
      
                        request.post({
                          headers: {
                            'content-type' : 'application/json',
                            'Authorization': tokenuser
                        },
                          url: 'http://10.4.200.20:9999/aiot-alert/v1/user-alert-sms/insert',
                          body: jsondata
                        }, function(error, response, body){
                          if (!error && response.statusCode == 200) {
                            msg.payload = "Success";
                            node.send(msg);
                          } else { 
                            bodyjson = JSON.parse(body);
                            node.warn(bodyjson.errorMessage);
                          }
                        });
                      } else {
                        node.warn("Not a valid phone nummber!");
                      }

                  }
              } else {
                node.warn("Error get token");
              }
            });
            
        });
    }
    RED.nodes.registerType("mobisms",SMSNode);
}
